src/infra/core/src/ingest/push_ingest_service_impl.rs

```diff
@@ -0,0 +1,368 @@
+            .get_reader(args.ctx.clone(), conf, temp_path)
+            .await?;
+
+        let df = reader.read(&input_data_path).await?;
+
+        Ok(Some(df))
+    }
+
+    // TODO: This is needed because tokio::fs::copy refuses to work with special
+    // files
+    async fn copy_special_file(
+        source_path: &Path,
+        target_path: &Path,
+    ) -> Result<(), InternalError> {
+        let mut source = std::fs::File::open(source_path).int_err()?;
+        let mut target = std::fs::File::create(target_path).int_err()?;
+
+        tokio::task::spawn_blocking(move || -> Result<(), std::io::Error> {
+            use std::io::{Read, Write};
+            let mut buf = [0u8; 2048];
+            loop {
+                let read = source.read(&mut buf)?;
+                if read == 0 {
+                    break;
+                }
+                target.write_all(&buf[..read])?;
+            }
+            Ok(())
+        })
+        .await
+        .int_err()?
+        .int_err()
+    }
+
+    async fn copy_stream_to_file(
+        mut data: Box<dyn AsyncRead + Send + Unpin>,
+        target_path: &Path,
+    ) -> Result<(), std::io::Error> {
+        use tokio::io::{AsyncReadExt, AsyncWriteExt};
+
+        let mut file = tokio::fs::File::create(target_path).await?;
+        let mut buf = [0u8; 2048];
+        loop {
+            let read = data.read(&mut buf).await?;
+            if read == 0 {
+                break;
+            }
+            file.write_all(&buf[..read]).await?;
+        }
+        file.flush().await?;
+        Ok(())
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+impl PushIngestService for PushIngestServiceImpl {
+    #[tracing::instrument(level = "info", skip_all, fields(%dataset_ref, %url, ?media_type))]
+    async fn ingest_from_url(
+        &self,
+        dataset_ref: &DatasetRef,
+        url: url::Url,
+        media_type: Option<MediaType>,
+        listener: Option<Arc<dyn PushIngestListener>>,
+    ) -> Result<PushIngestResult, PushIngestError> {
+        let listener = listener.unwrap_or_else(|| Arc::new(NullPushIngestListener));
+
+        self.do_ingest(dataset_ref, url, media_type, listener).await
+    }
+
+    #[tracing::instrument(level = "info", skip_all, fields(%dataset_ref, ?media_type))]
+    async fn ingest_from_file_stream(
+        &self,
+        dataset_ref: &DatasetRef,
+        data: Box<dyn AsyncRead + Send + Unpin>,
+        media_type: Option<MediaType>,
+        listener: Option<Arc<dyn PushIngestListener>>,
+    ) -> Result<PushIngestResult, PushIngestError> {
+        // Save stream to a file in cache
+        //
+        // TODO: Breaking all architecture layers here - need to extract cache into a
+        // service
+        let path = self
+            .cache_dir
+            .join(ingest_common::get_random_cache_key("push-ingest-"));
+        Self::copy_stream_to_file(data, &path).await.int_err()?;
```

Comment:
So, technically speaking, it's not exactly cached, but just stored in the cache directory? Maybe there should be a different temp directory?

Reply:
I removed the dependency on `cache_dir` and no longer abusing the cache.

I'm now passing the stream down into the body of the ingest flow, and similarly to `STDIN` the stream will be saved into a temporary `operation_dir` directory.